---
title: "Cflo-healthy-controls (TC7)"
author: Biplabendu Das
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    theme: united
    keep_md: no
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = F)
knitr::opts_chunk$set(message = F)

## For more inspiration on customizing the html output, refer to the following:
# https://bookdown.org/yihui/rmarkdown/html-document.html#table-of-contents

```

```{r housekeeping, include=FALSE}
set.seed(420)
rm(list = ls())

#' Load the libraries
pacman::p_load(pheatmap, dendextend, tidyverse, viridis)
pacman::p_load(RSQLite, tidyverse, dbplyr, DT, conflicted, WGCNA, igraph)
pacman::p_load(patchwork)
pacman::p_load(GeneOverlap)

#' set conflict preference
conflict_prefer("select", "dplyr", quiet = T)
conflict_prefer("filter", "dplyr", quiet = T)
conflict_prefer("union", "dplyr", quiet = T)
conflict_prefer("layout", "plotly", quiet = T)
conflict_prefer("hclust", "flashClust", quiet = T)

#' set path to your working directory
path_to_repo = "/Users/biplabendudas/Documents/GitHub/Das_et_al_2022b"

# script name
script.name = "02_build_annotate_circadianGCN"

#' load functions
# customized theme for publication quality figures
source(paste0(path_to_repo,"/functions/theme_publication.R"))
# function to perform enrichment analysis
source(paste0(path_to_repo,"/functions/enrichment_analysis.R"))
# function to plot z-scores (Cflo genes only)
source(paste0(path_to_repo,"/functions/plot_zscores.R"))

```


## Overview/Goals

Using time-course RNASeq data from healthy Cflo forager heads: 

- build a circadian gene co-expression network (GCN), 
- annotate the network using published data,

## Step 1: Build circadian GCN

### 1.1 Load data

Dataset: Healthy (uninfected, uninjected) Cflo forager ant heads (three pooled per time point for RNA-extraction and -sequencing), collected every 2h, over a 24h-period. [Control]

```{r load_data}
# Specify the path to TC5 repo
path_to_tc5_repo = "/Users/biplabendudas/Documents/GitHub/Das_et_al_2021"
# Load the TC5 database
tc5.db <- dbConnect(RSQLite::SQLite(), paste0(path_to_tc5_repo,"/data/TC5_data.db"))

# loading database which contains gene X sample expression data 
data.db <- dbConnect(RSQLite::SQLite(), paste0(path_to_repo,"/data/databases/TC7_data.db"))

# specify sample name
sample.name <- c("cflo_control","cflo_ophio-infected","cflo_beau-infected")

# extract the (gene-expr X time-point-of-sampling) data
dat <-
  data.db %>%
  tbl(., paste0(sample.name[1] ,"_fpkm")) %>%
  select(gene_name, everything()) %>%
  collect()

writeLines("What is the dimensions of the original dataset? [Rows = #genes, Cols = #samples]")
dim(dat[-1])


## Specify a col.scheme
col.scheme <- c("black","#AD212F","#5A829F", "#736D51", "#005C53")
names(col.scheme) <- c("control", "ophio-inf", "beau-inf", "control-brains", "control-brains-nur")

```

### 1.2 Clean data

The above dataset contains all genes (n=13,808) in the ant genome. 
However, not all of these genes are expressed in the ant head, and some are expressed at very low levels that are not biologically meaningful. Therefore, we will only keep the genes that are "expressed" (≥1 FPKM for at least half of all time points) in the ant head.

```{r clean_data}
# Which genes are expressed throughout the day in forager heads?
  # count the number of time points that has ≥ 1 FPKM
  n.expressed <- apply(dat[-1], 1, function(x) sum(x >= 1))
  # subset the data and only keep the filtered genes
  dat <- dat[which(n.expressed >=6),]

writeLines("Dimensions of the data post-filtering step [Rows = #genes, Cols = #samples]")
dim(dat)

# save the number of genes and samples
# that will be used to create the circadian GCN
nGenes = nrow(dat)
nSamples = ncol(dat[-1])

## - ## - ## - ## - ## - ## - ## - ## - ## - ## - ## - ## - 
# COME BACK HERE!!! ## - ## - ## - ## - ## - ## - ## - ## - 
## - ## - ## - ## - ## - ## - ## - ## - ## - ## - ## - ## - 

## Filter this data to keep only genes from specific modules
which.modules <- c("C2")

# module identity for Cflo-control gene
## specify the location of the csv that has this info.
file_loc <- "/results/WGCNA/cflo/cflo_heads_control_module_identity_new_labels.csv"
## load it into R
mod.identity.all <- read.csv(paste0(path_to_repo,file_loc), stringsAsFactors = F) %>% as_tibble()

## filter to keep only the genes that we are working with
mod.identity <-
  mod.identity.all %>%
  filter(module_identity %in% which.modules) %>% 
  select(gene_name, 
         module_identity) %>%
  # !!this step is necessary!! #
  arrange(module_identity)  


dat <- 
  dat %>% 
  filter(gene_name %in% mod.identity[["gene_name"]])

```

This is our cleaned, input data file for building the circadian GCN. 


### 1.3 Format data

- Log2 transform the data

```{r format_data}

datExpr = as.data.frame(t(log2(dat[-c(1)]+1)))
names(datExpr) = dat$gene_name
rownames(datExpr) = names(dat)[-c(1)]

# USE THE FOLLOWING CODE TO CHECK IF YOU HAVE ANY BAD SAMPLES #
  # gsg = goodSamplesGenes(datExpr, verbose = 3);
  # gsg$allOK

  # sampleTree = hclust(dist(datExpr0), method = "average");
  # # Plot the sample tree: Open a graphic output window of size 12 by 9 inches
  # # The user should change the dimensions if the window is too large or too small.
  # sizeGrWindow(12,9)
  # #pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
  # par(cex = 1);
  # par(mar = c(0,4,2,0))
  # plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5,
  #      cex.axis = 1.5, cex.main = 2)



# visualize the log-transformed data
x = reshape2::melt(as.matrix(t(datExpr)))
colnames(x) = c('gene_id', 'sample', 'value')

writeLines("Visualizing the log-transformed data")
ggplot(x, aes(x=value, col=sample)) + geom_density(size=2, alpha=0.8) + theme_Publication() + scale_colour_viridis_d()

```

### 1.4 Calculate gene-gene similarity

```{r gene_sim_matrix}
# Calculate Kendall's tau-b correlation for each gene-gene pair

# sim_matrix <- cor((datExpr), method = "kendall") # this step takes time
# save(sim_matrix, file = paste0(path_to_repo, "/results/temp_files/sim_matrix_", sample.name[1], "_part2_", 
#                                "_TC7.RData")) # might be useful to save the sim_matrix and
load(paste0(path_to_repo, "/results/temp_files/sim_matrix_", sample.name[1], "_part2_", "_TC7.RData")) # load it up

## Let's display a chunk of the matrix (code from Hughitt 2016; github)
heatmap_indices <- sample(nrow(sim_matrix), 200)

writeLines(paste0("Plotting a chunk of the gene-gene similarity matrix with ", length(heatmap_indices), " genes."))
gplots::heatmap.2(t(sim_matrix[heatmap_indices, heatmap_indices]),
          col=inferno(100),
          labRow=NA, labCol=NA,
          trace='none', dendrogram='row',
          xlab='Gene', ylab='Gene',
          main= paste0("Similarity matrix \n correlation method = 'kendall' \n (", length(heatmap_indices), " random genes)"),
          density.info='none', revC=TRUE)

```

### 1.5 Create adjacency matrix

- To create the adjacency matrix, we need to first identify the soft-thresholding power (see WGCNA for more info).

```{r soft_thresholding_power}
writeLines("Performing network topology analysis to pick soft-thresholding power")
# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to=30, by=2))
# # Call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)

writeLines("Plotting the resutls from the network topology analysis")
# Plot the results:
# sizeGrWindow(9, 5)
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col="red")
abline(h=0.75,col="darkgreen")
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")

```

> NOTE: The scale-free topology fit index reaches ~0.75 at a soft-thresholding-power=12, and it does not improve drastically beyond that.

Now, we can go ahead and create our adjacency matrix by power-transforming the similarity matrix (see WGCNA for more info).

```{r adjacency_matrix}
## Specify the soft-thresholding-power
soft.power = 12

# # # Construct adjacency matrix
# adj_matrix <- adjacency.fromSimilarity(sim_matrix,
#                                        power=soft.power,
#                                        type='signed')
# save(adj_matrix, file = paste0(path_to_repo, "/results/temp_files/adj_matrix_", sample.name[1], "_part2_",
#                                "_TC7.RData")) # might be useful to save the sim_matrix and
load(paste0(path_to_repo, "/results/temp_files/adj_matrix_", sample.name[1], "_part2_", "_TC7.RData")) # load it up


# Convert adj_matrix to matrix
gene_ids <- rownames(adj_matrix)

adj_matrix <- matrix(adj_matrix, nrow=nrow(adj_matrix))
rownames(adj_matrix) <- gene_ids
colnames(adj_matrix) <- gene_ids

writeLines(paste0("Plotting the power-transformed adjacency matrix for the same ", length(heatmap_indices)," genes as above"))
## Same heatmap as before, but now with the power-transformed adjacency matrix
gplots::heatmap.2(t(adj_matrix[heatmap_indices, heatmap_indices]),
                  col=inferno(100),
                  labRow=NA, labCol=NA,
                  trace='none', dendrogram='row',
                  xlab='Gene', ylab='Gene',
                  main='Adjacency matrix',
                  density.info='none', revC=TRUE)

## Delete similarity matrix to free up memory
rm(sim_matrix)
# gc()

```

***

## Step 2: Identify gene clusters

The following steps are performed as per guidelines from the WGCNA package and several tutorials made available online.

### 2.1 Create topological overalp matrix

```{r adj_to_TOM}
# # Turn adjacency into topological overlap
# TOM = TOMsimilarity(adj_matrix);
# dissTOM = 1-TOM
# save(dissTOM, file = paste0(path_to_repo, "/results/temp_files/dissTOM_", sample.name[1], "_part2_", 
#                             "_TC7.RData")) # might be useful to save the sim_matrix and
load(paste0(path_to_repo, "/results/temp_files/dissTOM_", sample.name[1], "_part2_", "_TC7.RData")) # load it up

# Call the hierarchical clustering function
geneTree = hclust(as.dist(dissTOM), method = "average")

writeLines("Plotting the resulting clustering tree (dendrogram)")
# sizeGrWindow(12,9)
# reset plotting parameter
par(mfrow = c(1,1))
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
     labels = FALSE, hang = 0.04)

```

### 2.2 Identify clusters

User defined parameters:

- minimum size (number of genes) of modules | var-name: minModuleSize

```{r identify_clusters}
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 50;

# Module identification using dynamic tree cut:
dynamicMods= cutreeDynamic(dendro = geneTree,
                           distM = dissTOM,
                           method = "hybrid",
                           verbose = 4,
                           deepSplit = 3, # see WGCNA for more info on tuning parameters
                           pamRespectsDendro = FALSE,
                           minClusterSize = minModuleSize);

# view number of genes in each module
# table(dynamicMods)

writeLines("How many genes are there in each of the initial modules (clusters) detected?
Note: The names of the modules (colors) have no meaning.")
# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)

```

### 2.3 Merge similar modules

User defined parameters:

- minimum correlation between two modules above which they are merged into one | var-name: MEDissThres


```{r refine_cluster}
# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes

# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs, method = "kendall");

writeLines("Clustering the module eigengenes and identifying a cutoff to merge similar modules")
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
# sizeGrWindow(7, 8)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "MEDiss = 1-cor(MEs, method = 'kendall')")

# We choose a height cut of 0.4, corresponding to correlation of 0.6, to merge
MEDissThres = 0.1 # user-specified parameter value; see WGCNA manual for more info

# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")

writeLines(paste0("Merging modules that have a correlation ≥ ", 1-MEDissThres))
# Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)

# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;

writeLines("Plotting the identified clusters (denoted with colors) before and after merging.")
# sizeGrWindow(12, 9)
plotDendroAndColors(geneTree,
                    cbind(dynamicColors, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)

# Rename to moduleColors
moduleColors = mergedColors

# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1

```

### 2.4 Calculate module-module similarity

```{r module_sim_matrix}
writeLines("Calculating module-module similarity based on module-eigengene-expression.")
# Calculate similarity of the eigen-genes
sim_matrix_ME <- cor(mergedMEs, method = "kendall")

# calculate adj_matrix
adj_matrix_ME <- adjacency.fromSimilarity(sim_matrix_ME,
                                          power=1, # DO NOT power transform
                                          type='signed'
)

## CHANGE THE NAMES OF THE MODULES;
module_ids <- data.frame(
  old_labels = rownames(adj_matrix_ME) %>% str_split("ME", 2) %>% sapply("[", 2) %>% as.character(),
  new_labels = paste0("S_", 1:nrow(adj_matrix_ME)))
# and coerce into matrix
adj_matrix_ME <- matrix(adj_matrix_ME, nrow=nrow(adj_matrix_ME))
rownames(adj_matrix_ME) <- module_ids$new_labels
colnames(adj_matrix_ME) <- module_ids$new_labels

# ## KEEP THE SAME MODULE NAMES (named by color)
# gene_ids <- rownames(adj_matrix_ME)
# # coerce into a matrix
# adj_matrix_ME <- matrix(adj_matrix_ME, nrow=nrow(adj_matrix_ME))
# rownames(adj_matrix_ME) <- gene_ids
# colnames(adj_matrix_ME) <- gene_ids

writeLines("Plotting the adjacency matrix that shows module-module similarity in expression")
gplots::heatmap.2(t(adj_matrix_ME),
                  col=inferno(100),
                  # labRow=NA, labCol=NA,
                  trace='none', dendrogram='row',
                  xlab='', ylab='',
                  # main='Similarity matrix - MEs \n correlation method = "kendall")',
                  main='Adjacency matrix - MEs \n modified edge weights)',
                  density.info='none', revC=TRUE)


```

### Calculate network stats

```{r global_network_stats}

global.network.stats <- qgraph::centrality_auto(adj_matrix_ME, 
                                        weighted = TRUE, signed = F)

```

Note: Nothing pops out from analyzing the centrality of different modules in the global network.
### 2.5 Visualize the network

```{r visualize_network}
pacman::p_load(igraph)

adj_matrix_ME_igraph <- adj_matrix_ME

# get rid of low correlations (0.6 & 0.8 are arbitrary) [0.7 and 0.9]
adj_matrix_ME_igraph[adj_matrix_ME_igraph < 0.8] <- 0
adj_matrix_ME_igraph[adj_matrix_ME_igraph < 0.9 & adj_matrix_ME_igraph>0] <- 0.5
adj_matrix_ME_igraph[adj_matrix_ME_igraph >= 0.9] <- 1

# build_network
network <- graph.adjacency(adj_matrix_ME_igraph,
                           mode = "upper",
                           weighted = T,
                           diag = F)

# simplify network
network <- igraph::simplify(network)  # removes self-loops

# remove isolated vertices (keep only the nodes)
isolated <- which(degree(network)==0)
network <- igraph::delete.vertices(network, isolated)

# E(network)$width <- E(network)$weight + min(E(network)$weight) + 1 # offset=1

# colors <- as.character(module_ids$old_labels)
# V(network)$color <- colors
V(network)$color <- "white"

# genes_ME <- factor(moduleColors, levels=colors) %>% summary()
V(network)$size <- igraph::degree(network, mode = "all")*3
# V(network)$size <- log2(genes_ME)^1.3

V(network)$label.color <- "black"
V(network)$frame.color <- "black"

E(network)$width <- E(network)$weight^2*4
E(network)$color <- "black"

## highlight shortest paths between two vetices
short.path <- igraph::get.shortest.paths(network, "S_5", "S_15")
E(network, path = unlist(short.path[[1]]))$color <- col.scheme[2]
E(network, path = unlist(short.path[[1]]))$width <- E(network)$weight*8

writeLines("Visualizing a simplified representation of the circadian GCN, with and without labels")
par(mfrow = c(1,2))

## Circular layout
# png(paste0(path_to_repo, "/results/figures/", sample.name[1], "/", script.name,"/", sample.name[1],"_GCN_1.png"),
#     width = 20, height = 30, units = "cm", res = 1000)
# par(bg=NA)
plot(network,
     layout=layout.kamada.kawai,
       # layout=layout.fruchterman.reingold,
       # layout=layout.graphopt,
       # layout=layout_in_circle,
     vertex.label=NA
     # vertex.size=hub.score(network)$vector*30
     # vertex.shape="none"
)
# dev.off()

# png(paste0(path_to_repo, "/results/figures/", sample.name[1], "/", script.name,"/", sample.name[1],"_GCN_2.png"),
#     width = 20, height = 30, units = "cm", res = 600)
# par(bg=NA)
plot(network,
     size=20,
     layout=layout.kamada.kawai,
       # layout=layout.fruchterman.reingold
       # layout=layout.graphopt
       # layout=layout_in_circle,
     # vertex.label=NA
     # vertex.size=hub.score(network)$vector*30
     vertex.shape="none"
)
# dev.off()
par(mfrow = c(1,1))
```

## Step 3: Annotate the network

### Module-gene list

```{r module_genes}
# Make a list that returns gene names for a given cluster
module_genes <- list()

# Get the genes from each of the modules
for (i in 1:length(module_ids$old_labels)) {
  
  # which color
  mod.color = as.character(module_ids$old_labels[[i]])
  
  # subset
  module_genes[[i]] <- names(datExpr)[which(moduleColors==mod.color)]
  names(module_genes)[[i]] <- as.character(module_ids$new_labels[[i]])
}
# # check the result | works
# names(module_genes)
# module_genes['C22']

# [13 Dec 2021]
# Save a csv with the module identity information for all genes used in building the GCN
# make a dataframe with gene_name and module_identity
for (i in 1:length(module_genes)){
  if (i == 1){
    cflo.control.mods <- data.frame(gene_name = module_genes[[i]],
                                    module_identity = as.character(names(module_genes)[i]))
  }
  else{
   foo <- data.frame(gene_name = module_genes[[i]],
                                    module_identity = as.character(names(module_genes)[i]))
    cflo.control.mods <- rbind(cflo.control.mods, foo) 
  }
  
}

# # save the dataframe as a csv
# cflo.control.mods %>% 
#   left_join(module_ids, by = c("module_identity" = "new_labels")) %>% 
#   write.csv(.,
#             paste0(path_to_repo,
#                    "/results/WGCNA/cflo/cflo_heads_control_module_identity_new_labels.csv"),
#             row.names = F)
# # done.

# # save a copy with all the gene annotations
# # load Cflo gene annotations
# cflo_annots <- read.csv(paste0(path_to_repo,"/functions/func_data/cflo_annots.csv"),
#                             header=T, stringsAsFactors = F, na.strings = c(NA,""," "))
# cflo.control.mods %>%
#   left_join(cflo_annots, by="gene_name") %>% head()
#   write.csv(.,
#             paste0(path_to_repo,
#                    "/results/WGCNA/cflo/cflo_heads_control_module_identity_new_labels_annots.csv"),
#             row.names = F)
```


### Identify the sub-clusters

```{r comparison_5}

## CHECK FOR OVERLAP

list1 <- module_genes

list2 <- list()
for (i in 1:length(which.modules)) {
  list2[[i]] <- mod.identity %>% filter(module_identity == which.modules[[i]]) %>% pull(gene_name)
  names(list2)[[i]] <- which.modules[[i]]
}

intersections <- list()
k <- 1
for (i in 1:length(list2)) {
  # writeLines("#########################################")
  # writeLines(paste0("Which primary module: ", names(list2)[[i]]))
  # writeLines("#########################################")
  g1 <- list2[[i]]
  for (j in 1:length(list1)) {
    # writeLines(paste0("which sub-module: ", names(list1)[[j]]))
    g2 <- list1[[j]]
    n_g1_g2 <- intersect(g1, g2)
    # save it in a list
    intersections[[k]] <- n_g1_g2
    names(intersections)[[k]] <- paste0(names(list2)[[i]], "_", names(list1)[[j]])
    k <- k+1
    
    # # print summary
    # writeLines(paste0("Percent of sub-module overlapping with primary module? ", round(length(n_g1_g2)/length(g2)*100,2)))
    # writeLines("----------------")
  }
}

## Make this a supplementary figure:

data.frame(sub_module = sapply(list1, length) %>% names()) %>% 
  mutate(n_genes = sapply(list1, length)) %>% 
  mutate(in_C21 = intersections[1:length(list1)] %>% sapply(., length)) %>% 
  mutate(in_C22 = intersections[(length(list1)+1):(length(list1)*2)] %>% sapply(., length)) %>% 
  
  pivot_longer(cols=starts_with("in"),values_to = "n_overlapping", names_to = "is_gene") %>% 
  
  mutate(percent_overlap = n_overlapping/n_genes*100) %>% 
  mutate(primary_module = ifelse(percent_overlap >= 80, str_split_fixed(is_gene,"_",2)[,2], "unassigned")) %>% 
  arrange(primary_module) %>% 
  
  filter(primary_module!="unassigned") %>% 
  filter(n_genes >= 50) %>% 
  
  select(sub_module, n_genes, percent_overlap, primary_module) %>%
  arrange(primary_module)


```

### Load genes of interest

```{r load_goi}
# load the list of all genes of interest
load(file = paste0(path_to_repo,"/results/genes_of_interest/goi_list.RData"))


## Rhythmic genes
rhy24.control <- goi.list[[1]][[1]][[1]]
rhy24.ocflo <- goi.list[[1]][[2]][[1]]
rhy24.beau <- goi.list[[1]][[3]][[1]]

## Ultradian genes
rhy12.control <- goi.list[[1]][[1]][[2]]
rhy12.ocflo <- goi.list[[1]][[2]][[2]]
rhy12.beau <- goi.list[[1]][[3]][[2]]
rhy08.control <- goi.list[[1]][[1]][[3]]
rhy08.ocflo <- goi.list[[1]][[2]][[3]]
rhy08.beau <- goi.list[[1]][[3]][[3]]

## DRGs
for.brain.head.rhy24.cluster1 <- goi.list[[2]]
for24.nur8 <- goi.list[[6]][[1]]

## DEGs
ocflo.up <- goi.list[[3]][[1]]
ocflo.down <- goi.list[[3]][[2]]
beau.up <- goi.list[[4]][[1]]
beau.down <- goi.list[[4]][[2]]
for.up <- goi.list[[5]][[1]]
for.down <- goi.list[[5]][[2]]

# ###-###-###-###
#   deg.dat <- readRDS(file = paste0(path_to_repo,"/results/cflo/","cflo_inf_v_control_DEGs.Rds"))
#   ophio.down <- deg.dat[[1]] %>% filter(inf_v_control == "down") %>% pull(gene_name)
#   beau.up <- deg.dat[[2]] %>% filter(inf_v_control == "up") %>% pull(gene_name)
# ###-###-###-###
# ###-###-###-###
#   deg.dat <- readRDS(file = paste0(path_to_repo,"/results/cflo/","cflo_inf_v_control_DEGs.Rds"))
#   ophio.up <- deg.dat[[1]] %>% filter(inf_v_control == "up") %>% pull(gene_name)
#   beau.down <- deg.dat[[2]] %>% filter(inf_v_control == "down") %>% pull(gene_name)
# ###-###-###-###


```

### Where are my genes at?

```{r comparison_focused}
writeLines("#####################################################
How many genes are in each of my geneset of interest?
#####################################################")

## MAKE YOUR LIST OF GENES OF INTEREST ##

# LIST ONE - WGCNA modules
list1 <- module_genes
writeLines("List of interesting genes #1
----------------------------
Genes in each of the identified gene-clusters or modules")
sapply(list1, length)

## LIST TWO - rhythmic genes
list3 <- list(rhy24.control,
              # rhy24.ocflo,
              # rhy24.beau,
              
              # rhy12.control,
              # rhy12.ocflo,
              # rhy12.beau,
              
              # rhy08.control,
              # rhy08.ocflo,
              # rhy08.beau,
              
              for.brain.head.rhy24.cluster1,
              for24.nur8,
              
              # for.up,
              # for.down,
              ocflo.up,
              beau.down,
              ocflo.down,
              beau.up)
writeLines("List of interesting genes #2
----------------------------
Rhythmic genes in control Cflo heads")
names(list3) <- c("rhy24-controls",
                  # "rhy24-Ocflo",
                  # "rhy24-Beau",
                  
                  # "rhy12-controls",
                  # "rhy12-Ocflo",
                  # "rhy12-Beau",
                  # 
                  # "rhy08-controls",
                  # "rhy08-Ocflo",
                  # "rhy08-Beau",
                  
                  "brain-head-rhy24-cluster1",
                  "for24-nur8",
                  
                  # "for-UP",
                  # "for-DOWN",
                  "Ocflo-UP",
                  "Beau-DOWN",
                  "Ocflo-DOWN",
                  "Beau-UP")
sapply(list3, length)


## CHECK FOR OVERLAP

## make a GOM object
gom.1v3 <- newGOM(list1, list3,
       genome.size = nGenes)
png(paste0(path_to_repo, "/results/figures/", sample.name[1],"_part2_gom_1v3.png"),
    width = 20, height =30, units = "cm", res = 300)
drawHeatmap(gom.1v3,
              adj.p=T,
              cutoff=0.01,
              what="odds.ratio", 
              # what="Jaccard",
              log.scale = T,
              note.col = "orange")
trash <- dev.off()

 # writeLines("How many genes exactly are overlapping between the pairwise comparisons")
# getMatrix(gom.1v4, name = "intersection") %>% t()

writeLines("Visualizing the significant overlaps between your lists of interesting genes and the identified modules")

```

```{r plot_gom_1v3, echo = FALSE, out.width='65%', fig.show='hold', fig.align='center', fig.cap='Where are my genesets of interests?'}
knitr::include_graphics(paste0(path_to_repo, "/results/figures/", sample.name[1],"_part2_gom_1v3.png"))
```



***

### Sub-module: S15 (DRGs)

Rhythmic genes

```{r S15_degs}
module.name <- "S_3"

# all S15 genes
list1[[module.name]] %>% # n = 125 genes
  
  # # filter to keep only the rhythmic ones
  # intersect(., rhy24.control) %>% # n = 76
  
  # ## deg plot
  # deg.plot_tc7()
  
  # enrichment
  timecourseRnaseq::check_enrichment(.,
                                     # what = "GOs",
                                     what = "pfams",
                                     org = "cflo",
                                     bg = mod.identity.all[["gene_name"]],
                                     filter = T,
                                     plot = T,
                                     expand=T) %>% view()
  
  # # zplots
  # stacked.zplot_tc7(conds="all", tc5 = F, bg.lwd = 0.5, alpha=1) %>%
  # multi.plot(rows = 3, cols=1)

  # # amplitude plots
  # amplitude.plot_tc7(cond = "all", tc5 = F, stats = T)

```

For links to circadian entrainment via hypoxic or oxidative stress in diseases, see https://www.frontiersin.org/articles/10.3389/fphys.2019.00682/full

https://www.nature.com/articles/s41467-021-25942-4

### Rhythmic genes in S14 and S15

Functions of these genes?

```{r S15_S14_degs}
module.name <- c("S_14","S_15")
# print(module.name)
# all S15 genes
list1[[module.name[[1]]]] %>% # n = 300 genes
  
  # combine the 14 and 15 together
  c(list1[[module.name[[2]]]]) %>% # n = 425
  
  # filter to keep only the rhythmic ones
  intersect(., rhy24.control) %>% # n = 122
  
  # # deg plot
  # deg.plot_tc7(
  #   # ylim=c(-1,1), 
  #   stats = T)
  
  # enrichment
  timecourseRnaseq::check_enrichment(.,
                                     what = "GOs",
                                     # what = "pfams",
                                     org = "cflo",
                                     bg = mod.identity.all[["gene_name"]],
                                     filter = T,
                                     plot = T,
                                     expand=T)
  

list1[[module.name[[1]]]] %>% # n = 300 genes
  
  # combine the 14 and 15 together
  c(list1[[module.name[[2]]]]) %>% # n = 425
  
  # filter to keep only the rhythmic ones
  intersect(., rhy24.control) %>% # n = 122
  # zplots
  stacked.zplot_tc7(conds="all", tc5 = F, plot.mean = T, bg.alpha = 0.2) %>%
  multi.plot(rows = 3, cols=1)

  # # amplitude plots
  # amplitude.plot_tc7(cond = "all", tc5 = T, stats = T)

```


### Sub-module: S5 (DEG)

Behavioral plasticity, task specialization

```{r S5_S3_degs}
module.name <- c("S_5","S_3")

# all s5 genes
list1[[module.name[[1]]]] %>% # n = 321 genes
  
  # # combine the s5 and s3 together
  c(list1[[module.name[[2]]]]) %>% # n = 553
  
  # # filter to keep only the rhythmic ones
  # intersect(., rhy.24[[1]]) %>% # n = 31 (s5), 40 (s5+s3)

  # # # Ocflo-UP-beau-DOWN
  # intersect(., ophio.up.beau.down) %>% 
  # TC7_annotator() %>% view()
  
  # # deg plot
  # deg.plot_tc7(ylim=c(-1,1), stats = T)
  
  # enrichment
  timecourseRnaseq::check_enrichment(.,
                                     what = "GOs",
                                     # what = "pfams",
                                     # what = "TMHMM",
                                     # what = "signalP",
                                     org = "cflo",
                                     bg = mod.identity.all[["gene_name"]],
                                     filter = T,
                                     plot = T,
                                     expand=T) 
  
  # # zplots
  # stacked.zplot_tc7(conds="all", tc5 = T) %>%
  # multi.plot(rows = 5, cols=1)

  # # amplitude plots
  # amplitude.plot_tc7(cond = "all", tc5 = T, stats = T)


```

Thioredoxins as the potential target for host clock hijacking?
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3772198/
https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003788 (thioredoxin reductase, beta-lactamases)

Thioredoxins in honeybees
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1847502/

TMX3
https://www.pnas.org/doi/10.1073/pnas.2003667117

